src/liballoc/boxed.rs

```diff
@@ -63,6 +63,27 @@
 //! T` obtained from `Box::<T>::into_raw` may be deallocated using the
 //! [`Global`] allocator with `Layout::for_value(&*value)`.
 //!
+//! `Box<T>` has the same ABI as `&mut T`. In particular, when `T: Sized`,
```

Comment:
Maybe something like:

---

So long as `T: Sized`, a `Box<T>` is guaranteed to be represented as a single pointer and is also ABI-compatible with C pointers (i.e., the C type `T*`). This means that you have Rust code which passes ownership of a `Box<T>` to C code by using `Box<T>` as the type on the Rust side, and `T*` as the corresponding type on the C side. As an example, consider this C header, which declares functions that create/return some kind of `Foo` value:

...

These two functions might be implemented in Rust as follows. Here, the `struct Foo*` types from `C` are translated to `Box<Foo>`, which captures the ownership constraints. Note also that the nullable argument to `destroy_foo` is represented in Rust as `Option<Box<Foo>>`, since `Box<Foo>` cannot be null.

...

Even though `Box<T>` has the same representation and C ABI as a C pointer, this does not mean that you can convert an arbitrary `T*` into a `Box<T>` and expect things to work. `Box<T>` values will always be fully aligned, non-null pointers. Moreover, the destructor for `Box<T>` will attempt to free the value with the global allocator. In general the best practice is to only use `Box<T>` for pointers that originated from the global allocator.

Reply:
@nikomatsakis No problem for the delay and thank you for the thorough review.

[The original commit](https://github.com/rust-lang/rust/pull/62514/commits/318c5d6c963d0e5d8ece89e804b4e696c6011955) that started this PR didn't mention references and focused on C pointers, but previous reviews recommended changing that.

I integrated your wording (slightly edited). Please let me know if that's ok. Thank you.